src/rcv/septentrio.c

```diff
@@ -119,802 +337,2401 @@ static uint8_t sig_tbl[SBF_MAXSIG+1][2]={ /* system, obs-code */
+    eph_t eph = {0};
+    int sat, svid, prn, ret;
+    uint8_t _buf[30] = {0};
+    int i = 0;
+
+    trace(3, "SBF decode_gpsrawcanav: len=%d\n", raw->len);
+
+    if (raw->len < 60) {
+        trace(2, "SBF decode_gpsrawcanav block length error: len=%d\n", raw->len);
         return -1;
     }
-    svid=U1(p);
-    if (!(sat = svid2sat(svid))||satsys(sat, &prn) != sys) {
-        trace(2,"sbf rawca svid error: sys=%d svid=%d\n", sys, svid);
+
+    /* get GPS satellite number */
+    svid = U1(p+8);
+    if (!(sat = svid2sat(svid)) || satsys(sat, &prn) != sys) {
+        trace(2, "sbf rawca svid error: sys=%d svid=%d\n", sys, svid);
         return -1;
     }
-    if (!U1(p+1)) {
+
+    if (!U1(p+9)) {
         trace(3, "sbf rawca parity/crc error: sys=%d prn=%d\n", sys, prn);
         return 0;
     }
+
     if (raw->outtype) {
-        sprintf(raw->msgtype+strlen(raw->msgtype)," prn=%d",prn);
+        if (sys == SYS_GPS)
+            sprintf(raw->msgtype, "SBF GPS Raw Navigation Data (PRN=%d)", prn);
+        if (sys == SYS_QZS)
+            sprintf(raw->msgtype, "SBF QZSS Raw Navigation Data (PRN=%d)", prn);
     }
-    for (i=0,p+=6;i<10;i++,p+=4) { /* 24 x 10 bits w/o parity */
-        setbitu(buff,24*i,24,U4(p)>>6);
+
+    /* clean up subframe from Septentrio. This is a little bit of work because
+     * Septentrio Rx add some parity bits to this message.
+     * We have to throw away the reserved bits as well as the parity bits.
+     */
+
+    /*   | 2bits |         24bits        |  6bits  |       <- SBF 32-bit word
+         ------------------------------------------
+                  | byte1 | bite2 | byte3 |                 <- sat nav message
+     */
+    for (i=0, p+=14; i<10; i++, p+=4) { /* 24 x 10 bits w/o parity */
+        setbitu(_buf, 24*i, 24, U4(p)>>6);
     }
-    id=getbitu(buff,43,3);
 
-    if (id<1||id>5) {
-         trace(2, "sbf rawca subframe id error: sys=%d prn=%d id=%d\n", sys, prn, id);
-         return -1;
-     }
-    memcpy(raw->subfrm[sat-1]+(id-1)*30,buff,30);
+    /* Now that we have a classic subframe we call the generic function */
+    id = getbitu(_buf, 43, 3); /* get subframe id */
+    if ((id<1) || (id>5)) {
+        trace(2, "sbf rawca subframe id error: sys=%d prn=%d id=%d\n", sys, prn, id);
+        return -1;
+    }
 
-    if (id==3) {
-        return decode_eph(raw,sat);
-     }
-     if (id==4||id==5) {
-         ret=decode_ionutc(raw,sat);
-         memset(raw->subfrm[sat-1]+id*30,0,30);
-         return ret;
-     }
-     return 0;
-}
-/* decode SBF GPS C/A subframe -----------------------------------------------*/
-static int decode_gpsrawca(raw_t *raw)
-{
-    return decode_rawca(raw,SYS_GPS);
+    memcpy(raw->subfrm[sat-1]+(id-1)*30, _buf, 30);
+
+    if (decode_frame(raw->subfrm[sat-1]   , &eph, NULL, NULL, NULL) == 1 &&
+        decode_frame(raw->subfrm[sat-1]+30, &eph, NULL, NULL, NULL) == 2 &&
+        decode_frame(raw->subfrm[sat-1]+60, &eph, NULL, NULL, NULL) == 3) {
+
```

Comment:
These uses of the return result of decode_frame() might have issues as decode_frame() returns only 1 for success and 0 for failure?

Reply:
Thanks. Will be changed in the next release.